src/v8/inspector/server.rs

```diff
@@ -0,0 +1,852 @@
+ * the Server Side Public License v1 (SSPLv1).
+ */
+//! This module provides with the debug server structures. With these
+//! structures it is possibly to easily start a debug server and start
+//! processing remote debuggers' messages over the network. The module
+//! implements messaging via the `WebSocket` protocol, as it is mainly
+//! used together with Chrome V8 engine for debugging, however, the
+//! implementation is platform-agnostic and one can use any sort of
+//! transport for the messages.
+//!
+//! # Clients
+//!
+//! To connect to a remote debugging WebSocket server, one can use:
+//!
+//! 1. A custom web-socket client, and some implementation of the
+//! V8 Inspector protocol.
+//! 2. The chrome-devtools, for example, by running the chromium
+//! web-browser and navigating to
+//! <devtools://devtools/bundled/inspector.html?ws=ip:port> where
+//! the `ip`/`port` parameters are the corresponding ip address and port
+//! of the [WebSocketServer] used.
+//! 3. (Needs verification) Using the WebKit MiniBrowser
+//! <https://trac.webkit.org/wiki/RemoteInspectorGTKandWPE>.
+//! 4. Using the V8's [D8 console debugger](https://v8.dev/docs/d8).
+//! 5. Using the [Visual Studio Code](https://code.visualstudio.com/),
+//! for example, using a configuration like this:
+//!     ```json
+//!     {
+//!         "version": "0.2.0",
+//!         "configurations": [
+//!             {
+//!                 "name": "Attach to RedisGears V8 on port 9005",
+//!                 "type": "node",
+//!                 "request": "attach",
+//!                 "cwd": "${workspaceFolder}",
+//!                 "websocketAddress": "ws://127.0.0.1:9005",
+//!             }
+//!         ]
+//!     }
+//!     ```
+//!
+//! # Example
+//!
+//! To debug, we must properly initialise the V8 and have a context
+//! scope ([super::V8ContextScope]), on which we can create an
+//! inspector.
+//!
+//! An inspector is just a hook into the `V8Platform` spying on the
+//! actions inside the [crate::v8::v8_context::V8Context] and which
+//! can pause, inspect and continue the execution.
+//!
+//! ```rust
+//! use v8_rs::v8::*;
+//! use inspector::messages::ClientMessage;
+//! use inspector::server::{DebuggerSession, TcpServer};
+//! use std::sync::{Arc, Mutex};
+//! use v8_rs::v8::inspector::RawInspector;
+//!
+//! // Initialise the V8 engine:
+//! v8_init_platform(1, Some("--expose-gc")).unwrap();
+//! v8_init().unwrap();
+//!
+//! // Create a new isolate:
+//! let isolate = isolate::V8Isolate::new();
+//!
+//! // Enter the isolate created:
+//! let i_scope = isolate.enter();
+//!
+//! // Create the code string object:
+//! let code_str = i_scope.new_string("1+1");
+//!
+//! // Create a JS execution context for code invocation:""
+//! let ctx = i_scope.new_context(None);
+//!
+//! // Enter the created execution context for debugging:
+//! let ctx_scope = ctx.enter(&i_scope);
+//!
+//! // Create an inspector.
+//! let inspector = Arc::new(RawInspector::new(&ctx_scope));
```

Comment:
Do you think we can come up with a simpler example that does not involve threading and locking? Also I am not sure why we need to wrap the inspector with Arc?

Reply:
Not follow, aren't we only simulate here the server? Why the server need to be multi threaded?